src/Components/Endpoints/src/Builder/RazorComponentEndpointDataSource.cs

```diff
@@ -139,9 +141,26 @@ private void UpdateEndpoints()
             _cancellationTokenSource = new CancellationTokenSource();
             _changeToken = new CancellationChangeToken(_cancellationTokenSource.Token);
             oldCancellationTokenSource?.Cancel();
+            oldCancellationTokenSource?.Dispose();
             if (_hotReloadService is { MetadataUpdateSupported : true })
             {
-                ChangeToken.OnChange(_hotReloadService.GetChangeToken, UpdateEndpoints);
+                if (_disposable != null)
+                {
+                    _disposable?.Dispose();
+                }
+                _disposable = ChangeToken.OnChange(_hotReloadService.GetChangeToken, UpdateEndpoints);
+            }
+        }
+    }
+ 
+    public void OnHotReloadClearCache(Type[]? types)
+    {
+        lock (_lock)
+        {
+            if (_disposable != null)
+            {
+                _disposable?.Dispose();
+                _disposable = null;
             }
```

Comment:
(because the `?.` already handles the null check)
```suggestion
            _disposable?.Dispose();
            _disposable = null;
```